compiler/src/iree/compiler/GlobalOptimization/PropagateDataLayout.cpp

```diff
@@ -0,0 +1,710 @@
+      return failure();
+    }
+
+    SetVector<Value> bestTransformedValues =
+        getTransformedValues(bestTf, costNodes, layoutID);
+    for (auto tf : barrierTransforms) {
+      SetVector<Value> transformedValues =
+          getTransformedValues(tf, costNodes, layoutID);
+      // Assume the cost of layout transformations are simple element-wise
+      // operations, so the cost of the transformations are proportional to the
+      // combined total number of elements in the set of transformed tensors.
+      if (hasMoreTotalElements(transformedValues, bestTransformedValues)) {
+        bestTf = tf;
+        bestTransformedValues = transformedValues;
+      }
+    }
+    LLVM_DEBUG({
+      llvm::dbgs() << "layoutID: " << layoutID << "\n";
+      llvm::dbgs() << "best tf: " << *bestTf << "\n\n";
+    });
+    if (bestTf->isIdentity())
+      return failure();
+    transform = bestTf;
+    return success();
+  }
+
+  LogicalResult getEdgeNodesAndBestLayoutTransformations(
+      SmallVector<SmallVector<Value>> &edgeNodes,
+      SmallVector<DataLayoutTransformation *> &transforms,
+      SmallVector<StringRef> &layoutIDs) {
+    DenseMap<StringRef, SetVector<GlobalDataLayoutValueElement *>> subgraphMap;
+    auto walkFn = [&](Value val) -> WalkResult {
+      if (auto *elementPtr =
+              solver.lookupElementFor<GlobalDataLayoutValueElement>(
+                  Position::forValue(val),
+                  /*queryingElement=*/nullptr, DFX::Resolution::NONE,
+                  /*allowInvalidState=*/false)) {
+        GlobalDataLayoutState &state = elementPtr->getState();
+        // Only support subgraphs with a single layout ID for now
+        auto stateLayoutIDs = state.getLayoutIDs();
+        if (stateLayoutIDs.size() != 1)
+          return WalkResult::advance();
+        if (subgraphMap.count(stateLayoutIDs[0])) {
+          subgraphMap[stateLayoutIDs[0]].insert(elementPtr);
+        } else {
+          SetVector<GlobalDataLayoutValueElement *> s;
+          s.insert(elementPtr);
+          subgraphMap.insert(std::make_pair(stateLayoutIDs[0], s));
+        }
+      }
+      return WalkResult::advance();
+    };
+    if (explorer.walkValues(walkFn) == TraversalResult::INCOMPLETE) {
+      return failure();
+    }
+
+    // For each subgraph, compute the optimal layout and endpoint nodes
+    for (auto [layoutID, valueElems] : subgraphMap) {
+      SmallVector<Value> subgraphEdgeNodes;
+      DataLayoutTransformation *subgraphTransform;
+      if (!failed(getSubgraphEdgeNodesAndBestLayoutTransformation(
+              valueElems, subgraphEdgeNodes, subgraphTransform, layoutID))) {
+        edgeNodes.push_back(subgraphEdgeNodes);
+        transforms.push_back(subgraphTransform);
+        layoutIDs.push_back(layoutID);
+      }
+    }
+    if (edgeNodes.empty()) {
+      return failure();
+    }
+    return success();
+  }
+
+  void annotateDataLayoutNodes() {
+    auto walkFn = [&](Value val) -> WalkResult {
```

Comment:
If you are not doing anything with `WalkResult` returned here, you can just make this a walk with void return type (and also remove the `WalkResult::advace()` below)